plugins/python/modulewrap.cpp

```diff
@@ -0,0 +1,632 @@
+    else if (inp_type == "long")
+      INSERT_INPUT_CONVERTER(long, inp);
+    else if (inp_type == "unsigned long")
+      INSERT_INPUT_CONVERTER(ulong, inp);
+    else if (inp_type == "float")
+      INSERT_INPUT_CONVERTER(float, inp);
+    else if (inp_type == "double")
+      INSERT_INPUT_CONVERTER(double, inp);
+    else if (inp_type.compare(0, 13, "numpy.ndarray") == 0) {
+      // TODO: these are hard-coded std::vector <-> numpy array mappings, which is
+      // way too simplistic for real use. It only exists for demonstration purposes,
+      // until we have an IDL
+      auto pos = inp_type.rfind("numpy.dtype");
+      if (pos == std::string::npos) {
+        PyErr_Format(
+          PyExc_TypeError, "could not determine dtype of input type \"%s\"", inp_type.c_str());
+        return nullptr;
+      }
+
+      pos += 18;
+
+      if (inp_type.compare(pos, std::string::npos, "int32]]") == 0) {
+        mod->ph_module->transform("pyvint_" + inp, vint_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else if (inp_type.compare(pos, std::string::npos, "uint32]]") == 0) {
+        mod->ph_module->transform("pyvuint_" + inp, vuint_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else if (inp_type.compare(pos, std::string::npos, "int64]]") == 0) { // need not be true
+        mod->ph_module->transform("pyvlong_" + inp, vlong_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else if (inp_type.compare(pos, std::string::npos, "uint64]]") == 0) { // id.
+        mod->ph_module->transform("pyvulong_" + inp, vulong_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else if (inp_type.compare(pos, std::string::npos, "float32]]") == 0) {
+        mod->ph_module->transform("pyvfloat_" + inp, vfloat_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else if (inp_type.compare(pos, std::string::npos, "double64]]") == 0) {
+        mod->ph_module->transform("pyvdouble_" + inp, vdouble_to_py, concurrency::serial)
+          .input_family(inp)
+          .output_products(inp + "py");
+      } else {
+        PyErr_Format(PyExc_TypeError, "unsupported array input type \"%s\"", inp_type.c_str());
+        return nullptr;
+      }
+    } else {
+      PyErr_Format(PyExc_TypeError, "unsupported input type \"%s\"", inp_type.c_str());
+      return nullptr;
+    }
+  }
+
+  // register Python algorithm
+  if (cinputs.size() == 1) {
+    if (!output.empty()) {
+      auto* pyc = new py_callback_1{callable}; // TODO: leaks, but has program lifetime
+      mod->ph_module->transform(cname, *pyc, concurrency::serial)
+        .input_family(cinputs[0] + "py")
+        .output_products("py" + output);
+    } else {
+      auto* pyc = new py_callback_1v{callable}; // id.
```

Comment:
It's not obvious that the `v` suffix here is for `void` and not `vector`.

Reply:
There's a greater split in use now b/c of the existence of separate transform and observe functions; I'll continue to see whether there's a hope of further cleanup, but these callbacks are now no longer used in the same function.